File: apps/agent-engine/app/schemas/match.py
"""Schemas cho tính năng Live Match — kết nối khách ↔ sale realtime qua Google Meet.

Mô hình "Uber cho tư vấn BĐS": khách online → hệ thống tìm sale online tốt nhất →
ping qua WebSocket (timeout 15s) → sale Accept → tạo Google Meet → cả hai vào call.

Toàn bộ schema dùng chung convention với app/schemas/* (Pydantic v2, Enum str,
datetime ISO8601). File này TÁCH RIÊNG khỏi CRM — không phụ thuộc app/schemas/crm.py.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel


class SaleAvailability(str, Enum):
    """Trạng thái sẵn sàng của sale trong hệ thống Live Match."""

    ONLINE = "online"  # sẵn sàng nhận khách
    BUSY = "busy"  # đang trong call
    AWAY = "away"  # offline / mất kết nối
    DO_NOT_DISTURB = "dnd"  # online nhưng không nhận match tự động


class MatchStatus(str, Enum):
    """Vòng đời một yêu cầu match."""

    PENDING = "pending"  # đang tìm sale
    INVITED = "invited"  # đã ping 1 sale, chờ accept (15s)
    ACCEPTED = "accepted"  # sale accept, đang tạo Meet
    LIVE = "live"  # Meet đã sẵn sàng, link gửi cho cả hai
    COMPLETED = "completed"  # call kết thúc (sale điền outcome)
    DECLINED = "declined"  # tất cả sale từ chối / không có sale
    EXPIRED = "expired"  # hết hạn invite mà không có ai accept
    CANCELLED = "cancelled"  # khách huỷ


class MatchOutcome(str, Enum):
    """Kết quả call do sale điền sau khi kết thúc (đổ về CRM ở Phase 2)."""

    INTERESTED = "interested"
    NOT_INTERESTED = "not_interested"
    BOOKED = "booked"  # đã chốt lịch xem nhà / đặt cọc
    FOLLOW_UP = "follow_up"  # cần gọi lại


class SalePresence(BaseModel):
    """Trạng thái hiện diện realtime của 1 sale (lưu in-memory ở presence.py)."""

    sale_id: str
    sale_name: str
    availability: SaleAvailability = SaleAvailability.ONLINE
    last_heartbeat_at: datetime
    active_calls: int = 0  # số call đang active
    last_match_at: Optional[datetime] = None  # cho round-robin fairness
    schedule_start: str = "08:00"  # giờ bắt đầu làm việc (giờ VN)
    schedule_end: str = "22:00"  # giờ kết thúc
    eligibility_score: float = 0.0  # điểm xếp hạng (từ sale_task_store)


class MatchRequest(BaseModel):
    """Một yêu cầu match đầy đủ (khớp record lưu match_store.json)."""

    id: str
    customer_id: str
    customer_name: str
    customer_email: str
    sale_id: Optional[str] = None
    sale_name: Optional[str] = None
    status: MatchStatus = MatchStatus.PENDING
    meet_link: Optional[str] = None
    meet_event_id: Optional[str] = None
    invited_sales: list[str] = []  # đã ping những sale nào
    declined_by: list[str] = []  # ai đã decline / expire
    invite_expires_at: Optional[datetime] = None  # hạn 15s của invite hiện tại
    created_at: datetime
    accepted_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    duration_seconds: Optional[int] = None
    outcome: Optional[MatchOutcome] = None
    outcome_note: Optional[str] = None


class CompleteMatchBody(BaseModel):
    """Body khi sale POST kết quả call (/match/{id}/complete)."""

    outcome: MatchOutcome
    note: Optional[str] = None


class MatchStats(BaseModel):
    """Thống kê tổng hợp cho dashboard admin."""

    period: str
    total: int = 0
    accepted: int = 0
    declined: int = 0
    expired: int = 0
    cancelled: int = 0
    live: int = 0
    completed: int = 0
    avg_duration_seconds: float = 0.0
    avg_accept_seconds: float = 0.0
    conversion_rate: float = 0.0  # completed / total
    online_sales: int = 0
    online_customers: int = 0
    active_calls: int = 0

File: apps/web/lib/match.ts
// Live Match — types + helper WebSocket/REST cho khách & sale.
// Khớp record backend (app/schemas/match.py). Tách riêng khỏi CRM.

import { AGENT_ENGINE_URL } from "@/lib/api";

export type MatchStatus =
  | "pending"
  | "invited"
  | "accepted"
  | "live"
  | "completed"
  | "declined"
  | "expired"
  | "cancelled";

export type MatchOutcome =
  | "interested"
  | "not_interested"
  | "booked"
  | "follow_up";

export type MatchRequest = {
  id: string;
  customer_id: string;
  customer_name: string;
  customer_email: string;
  sale_id: string | null;
  sale_name: string | null;
  status: MatchStatus;
  meet_link: string | null;
  meet_event_id: string | null;
  invited_sales: string[];
  declined_by: string[];
  invite_expires_at: string | null;
  created_at: string;
  accepted_at: string | null;
  completed_at: string | null;
  duration_seconds: number | null;
  outcome: MatchOutcome | null;
  outcome_note: string | null;
};

// Mọi message server push qua WebSocket.
export type WsMessage = {
  type: string;
  match?: MatchRequest;
  match_id?: string;
  meet_link?: string;
  expires_at?: string;
  timeout_seconds?: number;
  availability?: string;
  stats?: MatchStats;
  sale?: { id: string; name: string | null };
  fallback?: string;
  message?: string;
  [k: string]: unknown;
};

export type MatchStats = {
  period: string;
  total: number;
  accepted: number;
  declined: number;
  expired: number;
  cancelled: number;
  live: number;
  completed: number;
  avg_duration_seconds: number;
  avg_accept_seconds: number;
  conversion_rate: number;
  online_sales: number;
  online_customers: number;
  active_calls: number;
};

/** Base WS URL (http→ws, https→wss). */
export function wsBase(): string {
  return AGENT_ENGINE_URL.replace(/^http/i, "ws");
}

export function salePresenceWsUrl(token: string): string {
  return `${wsBase()}/ws/sale-presence?token=${encodeURIComponent(token)}`;
}

export function customerMatchWsUrl(token: string): string {
  return `${wsBase()}/ws/customer-match?token=${encodeURIComponent(token)}`;
}

function authHeaders(token: string): Record<string, string> {
  return { Authorization: `Bearer ${token}`, "Content-Type": "application/json" };
}

/** Sale POST kết quả sau call (REST). */
export async function completeMatch(
  token: string,
  matchId: string,
  outcome: MatchOutcome,
  note?: string,
): Promise<MatchRequest> {
  const res = await fetch(`${AGENT_ENGINE_URL}/match/${matchId}/complete`, {
    method: "POST",
    headers: authHeaders(token),
    body: JSON.stringify({ outcome, note: note ?? null }),
    cache: "no-store",
  });
  if (!res.ok) throw new Error(`Lỗi lưu kết quả (${res.status})`);
  return (await res.json()) as MatchRequest;
}

/** Khách huỷ match (REST fallback). */
export async function cancelMatch(token: string, matchId: string): Promise<void> {
  await fetch(`${AGENT_ENGINE_URL}/match/${matchId}/cancel`, {
    method: "POST",
    headers: authHeaders(token),
    cache: "no-store",
  }).catch(() => undefined);
}

/** Beep ngắn báo có khách (Web Audio — không cần file mp3, optional). */
export function playIncomingBeep(): void {
  if (typeof window === "undefined") return;
  try {
    const Ctx =
      window.AudioContext ||
      (window as unknown as { webkitAudioContext: typeof AudioContext })
        .webkitAudioContext;
    if (!Ctx) return;
    const ctx = new Ctx();
    const osc = ctx.createOscillator();
    const gain = ctx.createGain();
    osc.connect(gain);
    gain.connect(ctx.destination);
    osc.type = "sine";
    osc.frequency.value = 880;
    gain.gain.setValueAtTime(0.0001, ctx.currentTime);
    gain.gain.exponentialRampToValueAtTime(0.3, ctx.currentTime + 0.02);
    gain.gain.exponentialRampToValueAtTime(0.0001, ctx.currentTime + 0.5);
    osc.start();
    osc.stop(ctx.currentTime + 0.5);
    // đóng context sau khi phát để khỏi rò
    window.setTimeout(() => ctx.close().catch(() => undefined), 800);
  } catch {
    /* trình duyệt chặn autoplay — bỏ qua */
  }
}
